cip/1.accepted/CIP2017-04-20-query-combinators.adoc

```diff
@@ -0,0 +1,149 @@
+= CIP2017-04-20 - Query Combinators
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Author:* Stefan Plantikow <stefan.plantikow@neotechnology.com>
+
+[abstract]
+.Abstract
+--
+This CIP codifies the pre-existing `UNION` and `UNION ALL` clauses, and proposes additional query combinators for set operations and pipelining.
+--
+
+toc::[]
+
+== Motivation
+
+Query combinators for set operations are a common feature in other query languages.
+Adding more query combinators to Cypher will increase language expressivity and provide functionality that has been requested -- and expected to exist -- in the language by users.
+
+== Background
+
+The vast majority of Cypher clauses are underpinned by sequential composition; i.e. the records produced by the first clause act as an input to the next clause and so on.
+However, some operations require multiple streams of records as inputs.
+These are called _query combinators_.
+The most notable example of query combinators are _set operations_.
+
+== Proposal
+
+This CIP proposes the introduction of several new multi-arm query combinators:
+
+* `UNION`
+* `UNION MAX`
+* `UNION ALL`
+* `INTERSECT`
+* `INTERSECT ALL`
+* `EXCEPT`
+* `EXCEPT ALL`
+* `EXCLUSIVE UNION`
+* `EXCLUSIVE UNION ALL`
+* `OTHERWISE`
+* `CROSS`
+* `THEN`
+
+Multi-arm query combinators can only appear within a query using the syntax `<query> [<combinator> <query>]+`.
```

Comment:
I would prefer if the syntax was expressed in EBNF, similar to how other CIPs do that. 

```
query = singleQuery | combinedQuery ;
combinedQuery = singleQuery, queryCombinator, singleQuery ;
queryCombinator = union | intersect | except | otherwise | cross | then ;
union = "UNION", ["MAX" | "ALL"] ;
...
```